File: src/main/java/aurocosh/divinefavor/common/item/talismans/arrow/ArrowTalismanClimbableArrow.java
package aurocosh.divinefavor.common.item.talismans.arrow;

import aurocosh.divinefavor.common.entity.projectile.EntityClimbingArrow;
import aurocosh.divinefavor.common.entity.projectile.EntitySpellArrow;
import aurocosh.divinefavor.common.item.talismans.arrow.base.ArrowOptions;
import aurocosh.divinefavor.common.item.talismans.arrow.base.ArrowType;
import aurocosh.divinefavor.common.item.talismans.arrow.base.ItemArrowTalisman;
import aurocosh.divinefavor.common.spirit.base.ModSpirit;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.world.World;

import javax.annotation.Nonnull;
import java.awt.*;
import java.util.EnumSet;

public class ArrowTalismanClimbableArrow extends ItemArrowTalisman {
    protected final int despawnDelay;
    protected final float climbingSpeed;
    protected final float climbingDistance;
    protected final float climbingDistanceSq;

    public ArrowTalismanClimbableArrow(String name, ModSpirit spirit, int favorCost, Color color, double arrowDamage, EnumSet<ArrowOptions> options, ArrowType arrowType, float climbingSpeed, float climbingDistance, int despawnDelay) {
        super(name, spirit, favorCost, color, arrowDamage, options, arrowType);
        this.despawnDelay = despawnDelay;
        this.climbingSpeed = climbingSpeed;
        this.climbingDistance = climbingDistance;
        this.climbingDistanceSq = climbingDistance * climbingDistance;
    }

    @Override
    protected EntitySpellArrow getArrow(@Nonnull World world, EntityLivingBase shooter) {
        EntityClimbingArrow arrow = new EntityClimbingArrow(world, shooter);
        arrow.setClimbingStats(climbingSpeed, climbingDistanceSq, despawnDelay);
        return arrow;
    }

    public boolean onCollideWithPlayer(EntitySpellArrow spellArrow, EntityPlayer player) {
        return false;
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/item/talismans/arrow/ArrowTalismanClimbableArrow.kt
package aurocosh.divinefavor.common.item.talismans.arrow

import aurocosh.divinefavor.common.entity.projectile.EntityClimbingArrow
import aurocosh.divinefavor.common.entity.projectile.EntitySpellArrow
import aurocosh.divinefavor.common.item.talismans.arrow.base.ArrowOptions
import aurocosh.divinefavor.common.item.talismans.arrow.base.ArrowType
import aurocosh.divinefavor.common.item.talismans.arrow.base.ItemArrowTalisman
import aurocosh.divinefavor.common.spirit.base.ModSpirit
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.world.World
import java.awt.Color
import java.util.*

open class ArrowTalismanClimbableArrow(name: String, spirit: ModSpirit, favorCost: Int, color: Color, arrowDamage: Double, options: EnumSet<ArrowOptions>, arrowType: ArrowType, protected val climbingSpeed: Float, protected val climbingDistance: Float, protected val despawnDelay: Int) : ItemArrowTalisman(name, spirit, favorCost, color, arrowDamage, options, arrowType) {
    protected val climbingDistanceSq: Float = climbingDistance * climbingDistance

    override fun getArrow(world: World, shooter: EntityLivingBase): EntitySpellArrow {
        val arrow = EntityClimbingArrow(world, shooter)
        arrow.setClimbingStats(climbingSpeed, climbingDistanceSq, despawnDelay)
        return arrow
    }

    override fun onCollideWithPlayer(spellArrow: EntitySpellArrow, player: EntityPlayer): Boolean {
        return false
    }
}
